### DISCUSSION TITLE
type in atom(funciton)

### DISCUSSION THREAD

**Discussion by himself65** [pmndrs/jotai#1873] [Q&A] [answered]
I need to pass a function into an atom in my code structure. 

But I have to use a type cast ugly like this.

For example

```ts
type FetchApi = <T>(key: string) => Promise<T>;

export const fetchAtom = atom<FetchApi>(
  null! as FetchApi
);

defaultStore.set(fetchAtom, () => {
  return async (key: string): Promise<any> => {
    return new Promise<any>((resolve) => {
      setTimeout(() => resolve({ id: 1, name: "test" }), 1000);
    });
  };
});
```

Otherwise, the type of this atom will be `Atom`, instead of `PrimitiveAtom`

I want to know if I there's any improvement for kind of this code or design?

**Comment by dai-shi** [comment, marked as answer]
Storing a function as atom value is technically unsupported, because it doesn't work with function updates.
So, the recommendation is to wrap a function with an object.

```ts
export const fetchAtom = atom<{ fn: FetchApi }>({
  fn: null! as FetchApi
});
